Tour Chart Inverted Pace

FJBDev opened the pull request:
Currently, MyTourbook shows the pace graph with descending values : The faster the user is running, the lower the value is and the slower the user is running, the higher the value is.
This is not intuitive because it is essentially an "inverted" graph when we compare it to the speed graph and makes it harder for the user to "read" the graph.

![image](https://user-images.githubusercontent.com/25235665/109083473-36802480-76c3-11eb-8b2b-3edff7c0ffa4.png)


Example with Strava where the pace is correctly displayed:
![image](https://user-images.githubusercontent.com/25235665/109083112-86aab700-76c2-11eb-8cd4-73d4db5ba35a.png)

After the [fix](https://github.com/wolfgang-ch/mytourbook/pull/306/commits/c64090fcdcf6abcf8d426cc0f7d1c1d517471169):
![image](https://user-images.githubusercontent.com/25235665/109404948-62501400-7928-11eb-995f-5af71aafbedc.png)


- Misc code improvements found when I tested my code : 
  - [Array designators "[]" should be located after the type in method signatures](https://rules.sonarsource.com/java/tag/convention/RSPEC-1195)
  - [Removing unused parameters](https://rules.sonarsource.com/java/tag/unused/RSPEC-1172)
  - [Making anonymous inner class a lambda](https://rules.sonarsource.com/java/RSPEC-1604)

wolfgang-ch commented:
> @wolfgang-ch My code is not complete because I haven't been able to find how to display properly the Y-Axis legend in proper minute/mile from the speed serie. Any thoughts ? I am still very new to the TourChart code

The day time statistic contains an inverted y axis
https://github.com/wolfgang-ch/mytourbook/blob/c1a6ee3328e21904ab075aa69caf4a34ba172081/bundles/net.tourbook.statistics/src/net/tourbook/statistics/graphs/StatisticTour_Time.java#L450

There is a bug since every and nobody complained until now, when dragging the unit slider (horizontal axis) then the opposite slider is dragged.

FJBDev commented:
> The day time statistic contains an inverted y axis
> 
Actually, that's where I found the function `setYAxisDirection` but the day time is not "inverted" like the pace should be.

wolfgang-ch commented:
This method

https://github.com/wolfgang-ch/mytourbook/blob/c1a6ee3328e21904ab075aa69caf4a34ba172081/bundles/net.tourbook.chart/src/net/tourbook/chart/ChartComponentGraph.java#L2627

must be improved to support y-axis direction, similar like this

https://github.com/wolfgang-ch/mytourbook/blob/c1a6ee3328e21904ab075aa69caf4a34ba172081/bundles/net.tourbook.chart/src/net/tourbook/chart/ChartComponentGraph.java#L3641-L3647

but it is possible that this code is currently not used

FJBDev commented:
Thanks, I will investigate that code

FJBDev commented:
Converting to draft so you don't receive notification emails for each commit i will do

FJBDev commented on bundles/net.tourbook/src/net/tourbook/tour/TourManager.java:
```diff
@@ -4259,12 +4260,10 @@ private ChartDataYSerie createModelData_Pace(final TourData tourData,
          yDataPace.setSliderLabelFormat(ChartDataYSerie.SLIDER_LABEL_FORMAT_MM_SS);
          yDataPace.setCustomData(ChartDataYSerie.YDATA_INFO, GRAPH_PACE);
          yDataPace.setCustomData(CUSTOM_DATA_ANALYZER_INFO, new TourChartAnalyzerInfo(true, false, _computeAvg_Pace, 1));
+         yDataPace.setYAxisDirection(false);
```
@wolfgang-ch This is part of the fix

FJBDev commented on bundles/net.tourbook.chart/src/net/tourbook/chart/ChartComponentGraph.java:
```diff
@@ -2970,8 +2934,11 @@ private void drawAsync_510_LineGraphSegment(final GC gc,
 
-                  devY = devY0Inverse - devY1;
-
+                  if (yData.isYAxisDirection()) {
```
@wolfgang-ch This is the 2nd part of the fix.

The rest is code improvements I found when I modified those files.

wolfgang-ch commented on bundles/net.tourbook.chart/src/net/tourbook/chart/ChartComponentGraph.java:
```diff
@@ -3454,18 +3410,14 @@ private void drawAsync_530_BarGraph(final GC gcGraph, final GraphDrawingData dra
-               if (devXPosNextBar > 0) {
-                  if (devXPos < devXPosNextBar) {
+            if (serieLayout != ChartDataYSerie.BAR_LAYOUT_SINGLE_SERIE && devXPosNextBar > 0 && devXPos < devXPosNextBar) {
 
```
Why not keep a better readable code

![mt unformatted code](https://user-images.githubusercontent.com/1283445/109531443-2ba90380-7ab8-11eb-9c7d-e480f92c716c.png)

FJBDev replied:
Reverted

wolfgang-ch commented on bundles/net.tourbook.chart/src/net/tourbook/chart/ChartComponentGraph.java:
```diff
@@ -6364,13 +6310,10 @@ private ChartYSlider isYSliderHit(final int graphX, final int devY) {
                || (_canChartBeOverlapped && (isStackedChart || isLastOverlappedGraph));
 
-         if (canDoHitChecking) {
-
-            if (ySlider.getHitRectangle().contains(graphX, devY)) {
+         if (canDoHitChecking && ySlider.getHitRectangle().contains(graphX, devY)) {
 
-               _hitYSlider = ySlider;
-               return ySlider;
-            }
```
Sometimes this complex code in the tour chart is not "optimized" to do better debugging, e.g. you cannot set in the optimized code a breakpoint to the old line 6369, this mostly effects also ALL THE OTHER OPTIMIZATIONS

FJBDev replied:
Reverted

wolfgang-ch commented on bundles/net.tourbook.statistics/src/net/tourbook/statistics/graphs/StatisticDay.java:
```diff
@@ -188,91 +175,76 @@ public void createStatisticUI(final Composite parent, final IViewSite viewSite)
-      _tourToolTip.addHideListener(new IToolTipHideListener() {
-         @Override
-         public void afterHideToolTip(final Event event) {
-            // hide hovered image
```
Comment is gone

FJBDev replied:
Reverted

wolfgang-ch commented on bundles/net.tourbook.statistics/src/net/tourbook/statistics/graphs/StatisticTour_Time.java:
```diff
@@ -135,86 +128,71 @@ public void createStatisticUI(final Composite parent, final IViewSite viewSite)
-      _tourToolTip.addHideListener(new IToolTipHideListener() {
-         @Override
-         public void afterHideToolTip(final Event event) {
-            // hide hovered image
```
Comment is gone

FJBDev replied:
Reverted

wolfgang-ch commented on bundles/net.tourbook.statistics/src/net/tourbook/statistics/graphs/StatisticTraining.java:
```diff
@@ -189,83 +176,68 @@ public void createStatisticUI(final Composite parent, final IViewSite viewSite)
-      _tourToolTip.addHideListener(new IToolTipHideListener() {
-         @Override
-         public void afterHideToolTip(final Event event) {
-            // hide hovered image
```
Comment is gone

FJBDev replied:
Reverted

wolfgang-ch reviewed (commented):
The tour chart is a stable complex code, there was no need to modify tons of codelines just to implement a few new/modified lines for this feature. It took me an hour to review all the modifications instead of reviewing just a few lines of code.

![mt-tons-of modifications](https://user-images.githubusercontent.com/1283445/109533541-a2df9700-7aba-11eb-9e9c-a23d9ea529a4.png)

FJBDev commented:
> 
> 
> The tour chart is a stable complex code, there was no need to modify tons of codelines just to implement a few new/modified lines for this feature. It took me an hour to review all the modifications instead of reviewing just a few lines of code.
> 
> ![mt-tons-of modifications](https://user-images.githubusercontent.com/1283445/109533541-a2df9700-7aba-11eb-9e9c-a23d9ea529a4.png)

Ok, I reverted all my code improvements in this new [PR](https://github.com/wolfgang-ch/mytourbook/pull/310)

>there was no need to modify tons of codelines just to implement a few new/modified lines for this feature
There was no need but I like to follow this [standard practice](https://www.matheus.ro/2017/12/11/clean-code-boy-scout-rule/) : "Leave the code cleaner than you found it."
But I have removed those.

FJBDev commented on bundles/net.tourbook/src/net/tourbook/ui/tourChart/TourChart.java:
```diff
@@ -2487,35 +2462,31 @@ public void drawSelectedLines(final GC gc,
                   path1.lineTo(graphX1, graphY1);
 
-               } else if (pointIndex > 4) {
-
+               } else if (pointIndex > 4 && graphX1 != graphXPrev || graphY1 != graphYPrev || prevRGB != segmentRGB) {
                   // draw following segments
 
                   // optimize, draw only when changed
-                  if (graphX1 != graphXPrev || graphY1 != graphYPrev || prevRGB != segmentRGB) {
```
Reverted

FJBDev commented on bundles/net.tourbook/src/net/tourbook/ui/tourChart/TourChart.java:
```diff
@@ -1899,8 +1877,7 @@ private ChartLabel createLayer_Marker_ChartLabel(final TourMarker tourMarker,
     */
    private ChartLabel createLayer_Pause_ChartLabel(final String pauseDuration,
                                                    final double[] xAxisSerie,
-                                                   final int xAxisSerieIndex,
-                                                   final int labelPosition) {
```
Reverted

FJBDev commented on bundles/net.tourbook/src/net/tourbook/ui/tourChart/TourChart.java:
```diff
@@ -827,14 +818,7 @@ public void widgetDisposed(final DisposeEvent e) {
       _tourInfoIconTooltip = new TourToolTip(getToolTipControl());
       _tourInfoIconTooltip.addToolTipProvider(_tourInfoIconTooltipProvider);
 
-      _tourInfoIconTooltip.addHideListener(new IToolTipHideListener() {
-         @Override
-         public void afterHideToolTip(final Event event) {
-
-            // hide hovered image
-            getToolTipControl().afterHideToolTip(event);
-         }
-      });
```
Reverted

FJBDev commented on bundles/net.tourbook/src/net/tourbook/ui/tourChart/SlideoutGraphMinMax.java:
```diff
@@ -91,12 +90,9 @@ public void widgetSelected(final SelectionEvent e) {
-            UI.adjustSpinnerValueOnMouseScroll(event);
-            onChangeUI();
-         }
+      _defaultMouseWheelListener = mouseEvent -> {
+         UI.adjustSpinnerValueOnMouseScroll(mouseEvent);
+         onChangeUI();
```
Reverted

FJBDev commented on bundles/net.tourbook/src/net/tourbook/tour/TourManager.java:
```diff
@@ -4259,12 +4260,10 @@ private ChartDataYSerie createModelData_Pace(final TourData tourData,
          yDataPace.setCustomData(CUSTOM_DATA_ANALYZER_INFO, new TourChartAnalyzerInfo(true, false, _computeAvg_Pace, 1));
+         yDataPace.setYAxisDirection(false);
 
-         if (useGraphBgStyle) {
-            yDataPace.setGraphFillMethod(ChartDataYSerie.FILL_METHOD_CUSTOM);
-         } else {
-            yDataPace.setGraphFillMethod(ChartDataYSerie.FILL_METHOD_FILL_BOTTOM);
-         }
```
Reverted

wolfgang-ch commented:
> 
> 
> > The tour chart is a stable complex code, there was no need to modify tons of codelines just to implement a few new/modified lines for this feature. It took me an hour to review all the modifications instead of reviewing just a few lines of code.
> 
> Ok, I reverted all my code improvements in this new [PR](https://github.com/wolfgang-ch/mytourbook/pull/310)
 
> > there was no need to modify tons of codelines just to implement a few new/modified lines for this feature
>>
>  There was no need but I like to follow this [standard practice](https://www.matheus.ro/2017/12/11/clean-code-boy-scout-rule/) : "Leave the code cleaner than you found it."
>  But I have removed those.

The question is, what is clean code? I don't think it is e.g. removing comments, set comment to the wrong location, make the code less readable which are the findings in the issues.

Not all was bad what you had done, I already use also lamda listeners but it seems you have used the quick fix shortcuts or refactoring tool and not reviewed the code changes. There are refactoring tools in Eclipse, which I do not use because it creates a mess however you can mostly preview the changes.

FJBDev commented:
> The question is, what is clean code? I don't think it is e.g. removing comments, set comment to the wrong location, make the code less readable which are the findings in the issues.

I think the answer is different for each person. Just like for the question what is "readable code"

FJBDev commented:
> Not all was bad what you had done, I already use also lamda listeners but it seems you have used the quick fix shortcuts or refactoring tool and not reviewed the code changes. There are refactoring tools in Eclipse, which I do not use because it creates a mess however you can mostly preview the changes.

Ok, I will do again the lambda listeners but probably in a subsequent PR because actually, I prefer to have the code to fix the pace chart in a separate commit so that is obvious what needed to be changed.

wolfgang-ch commented:
> 
> 
> > Not all was bad what you had done, I already use also lamda listeners but it seems you have used the quick fix shortcuts or refactoring tool and not reviewed the code changes. There are refactoring tools in Eclipse, which I do not use because it creates a mess however you can mostly preview the changes.
> 
> Ok, I will do again the lambda listeners but probably in a subsequent PR because actually, I prefer to have the code to fix the pace chart in a separate commit so that is obvious what needed to be changed.

You should do it in a separate PR

FJBDev commented:
> You should do it in a separate PR

Yes, I will and that is what I meant by this "I prefer to have the code to fix the pace chart in a separate commit so that is obvious what needed to be changed."
